Implement extend to title for Windows and Linux

Hilderin opened the pull request:
- Implements proposal https://github.com/godotengine/godot-proposals/issues/4846

This PR implements the `EXTEND_TO_TITLE` flag for Windows and Linux X11. Implementing this flag allows Windows and Linux to use the editor without the default title bar. Additionally, it makes it possible to use this flag in any Godot project outside the editor.

Since Windows and Linux don’t natively support this mode, it is "simulated" by providing a resizable borderless window with minimize, maximize, and close buttons rendered directly in the `Window` class. New properties were added to the `Window Theme` to define the button icons, positions, and styles.

Here’s a little demo:

Windows:

https://github.com/user-attachments/assets/520e4e41-4682-43b8-b307-26b0521cd5b3

Linux:
![image](https://github.com/user-attachments/assets/d82882e0-0871-41bc-8d74-2c7a886387cb)

![image](https://github.com/user-attachments/assets/1c2df54c-d0e3-4a70-ac23-5ce2941dbe11)

Hilderin committed:
Editor without title bar

bruvzg commented on editor/editor_node.cpp:
```diff
@@ -1253,19 +1272,18 @@ void EditorNode::_viewport_resized() {
-		title_bar->set_custom_minimum_size(Size2(0, margin.z - title_bar->get_global_position().y));
-	}
+
+	int w = (gui_base->is_layout_rtl()) ? margin.y : margin.x;
+	left_menu_spacer->set_custom_minimum_size(Size2(w, 0));
+	right_menu_spacer->set_custom_minimum_size(Size2(w, 0));
```
Spacers are there to reserve space for native window close/min/max buttons on macOS. It should use opposite X/Y values for different sides, not a single W for both.

bruvzg commented:
I'm not sure if we want a separate flag for it, might be better to merge it with `FEATURE_EXTEND_TO_TITLE` (it's the same, but OS draw the buttons) and move window min/max/close buttons from a separate controls to `Window` class (for non macOS).

bughandler commented:
I've been working on the same feature for the past few days and have a few questions. 
1. I wonder if we really need to add a new flag instead of using the existing `WINDOW_FLAG_EXTEND_TO_TITLE`, which is already widely used in the MacOS implementation.
2. As shown in the screenshot below, I believe your implementation may not work on Windows 7/10 when the editor is in windowed mode, as it could result in a white border around the window. You may need to leverage the DWM APIs to resolve this issue.
3. Additionally, the project manager window should have the same no-title-bar appearance when the corresponding setting is enabled.

![image](https://github.com/user-attachments/assets/fc93df32-514e-4d19-8359-0da9b76815fc)

Hilderin commented:
> I'm not sure if we want a separate flag for it, might be better to merge it with FEATURE_EXTEND_TO_TITLE (it's the same, but OS draw the buttons) and move window min/max/close buttons from a separate controls to Window class (for non macOS).

That's a interesting suggestion, I'll think about it. It should be easier to implement in the Project Manager or other windows in the future if the buttons are drawn directly in the Window class. I thought it could be interesting for games/app created with Godot to be able to create resizable without without the window min/max/close buttons. It could be added later.

Hilderin commented:
> I wonder if we really need to add a new flag instead of using the existing WINDOW_FLAG_EXTEND_TO_TITLE, which is already widely used in the MacOS implementation.

I'll look into it, bruvzg made the same suggestion.

> As shown in the screenshot below, I believe your implementation may not work on Windows 7/10 when the editor is in windowed mode, as it could result in a white border around the window. You may need to leverage the DWM APIs to resolve this issue.

I'm curious how you implemented it.

> Additionally, the project manager window should have the same no-title-bar appearance when the corresponding setting is enabled.

Indeed.

bughandler commented:
> I'm curious how you implemented it.

FYI: <https://learn.microsoft.com/en-us/windows/win32/dwm/customframe>

Hilderin commented:
I fixed the issue with the white border on top of the window on Windows 7/10. Thanks bughandler!

If you want to work on that together @bughandler, I'm available in the Godot Contributors Chat: https://chat.godotengine.org

Hilderin commented:
@bruvzg I was working on using the already existing flag `WINDOW_FLAG_EXTEND_TO_TITLE` and implementing the rendering of the window buttons directly in the `Window` class. I have no problem using this flag, but I'm concerned about rendering the buttons directly in the `Window` class.

Should I create a control for that and add it to the scene? That would be the easiest way, but my concern is that these controls would be present in the scene tree at runtime for any project that uses `WINDOW_FLAG_EXTEND_TO_TITLE`.

The other alternative could be to manage the rendering manually, like in `Viewport::_sub_window_update`, but that seems to be very complex, especially for handling mouse over, clicks, etc.

What’s the best approach?

Hilderin commented:
Added support for Linux x11...

![image](https://github.com/user-attachments/assets/1fff4908-6271-49eb-bc1e-3531138ad825)

![image](https://github.com/user-attachments/assets/a90f0c5e-4e2e-4120-8ee5-39d4378b7b81)

Hilderin commented:
@Riteo I would certainly appreciate your contribution for the Wayland port!

For the CSDs, I'm currently tinkering with the idea to implement them directly into the Window class. I guess adding the title and options to move the window directly there for Wayland could be added afterward.

Hilderin commented:
After some work, I was able to put the rendering of the window buttons directly in the `Window` class. I added some properties to Window Theme to handle the styles and the position of the buttons. I reverted most of the modifications in `EditorNode` because Windows and Linux implementation are now closer to the macOS implementation. The method `window_is_extend_to_title_show_window_buttons` in `DisplayServer` is used to know if the windows buttons are rendered by the OS or if Godot needs to rendered them. Only `DisplayServerMacOS::window_is_extend_to_title_show_window_buttons` returns true.

Having a `decoration_canvas` now directly in the `Window` class could be used in the future to a full custom title bar for Wayland.

Here is a test project to test some Window settings at runtime:
[test-godot-window-different-types.zip](https://github.com/user-attachments/files/17007633/test-godot-window-different-types.zip)

geekley commented:
This is amazing!

I can't see in the video, but I'm assuming the windows in play mode / exported game are draggable as well. Are they? By the "title" area or by the entire window? Only on empty space?

---
I'd also like to repeat my suggestion from before.
This serves to not lose information in this mode, while still making the best use of the space:
> Add the icon and project name (the part that would show on the titlebar) at the top-left:
> You could use the project name to replace the "Project" menu and show it before the other menus

So, ideally the menu part on the top left could be like this:
🟦 `My Super Game ▼` | `Scene` | `Debug` | `Editor` | `Help`
Using "My Super Game" as the project name, and 🟦 to represent either a Godot icon or the Project's icon.

There's no need to repeat the scene name and `(*)` dirty state that are already on the tabs.
Joining Project name with the project menu makes total sense, as long as you add a ▼ symbol to make it obvious that you can click it as a menu. It's also better to have it before the `Scene` menu, and you can then move the out-of-place `Quit` command to this Project menu, right below `Quit to Project List` where it belongs.

Calinou reviewed (changes requested):
Tested locally on Windows 11 23H2, it works as expected.
This is a great start; it makes the editor feel a lot more polished already 🙂 

Some feedback:

- I find the minimize/maximize/close buttons a bit tall compared to other buttons on Windows 11 (each button has a 36x36 clickable area):

![image](https://github.com/user-attachments/assets/3bca3bb3-204c-49c4-b023-be932e1ac120)

Compare to VS Code's, which are also client-side decorations (each button has a 45x32 clickable area):

![image](https://github.com/user-attachments/assets/a28d0646-7e39-4876-86a0-e75695e0a615)

- When a window is maximized or fullscreen (and only when maximized), the minimize/maximize/close buttons should be placed **exactly** in the window corner with no vertical or horizontal margin, so that when the window is maximized, you can fly your cursor to the top-right corner and left-click to close the window. No need for precise aiming. This behavior also works in VS Code, for instance.
  - In the editor, the top menu buttons can't be clicked right away if you fly your mouse to the top due to similar margins:

![image](https://github.com/user-attachments/assets/c79e44aa-2798-4fc6-941c-ca0cc80c0145)

When the window is made fullscreen, the minimize/maximize/close buttons disappear entirely, but they still take invisible space, which looks odd:

![image](https://github.com/user-attachments/assets/323de7f5-b8ee-4573-b9c9-0b011d54dbef)

In comparison, the space is reclaimed by the buttons next to it in VS Code.

- Pressing the <kbd>Shift + F11</kbd> shortcut or using **Editor > Toggle Fullscreen** does not make the minimize/maximize/close buttons reappear, even though you can still click the area they previously occupied for the usual effect.

- Window borders are momentarily visible during the project manager splash screen. To avoid this visual discrepancy, we should make the splash screen borderless if **Expand to Title** is enabled in the Editor Settings.

KoBeWi commented:
Shouldn't buttons use the same color as background around them? Right now they oddly stand out.

Calinou commented:
> Shouldn't buttons use the same color as background around them? Right now they oddly stand out.

Yes, I think we could get rid of their background colors (when not hovered) too.

Hilderin commented:
In "Extend to ttitle" mode, the moving of the window when dragging the "title bar" is not done in the DisplayServer, it's done into the `EditorTitleBar`. Effectively, it's would be very difficult to know what is an "empty space" or not from the DisplayServer. If you are trying to implement a complete custom title bar while not in "Extend to title" mode, I guess you can catch the events in the rect corresponding to the size of the title bar.

Hilderin commented:
> I can't see in the video, but I'm assuming the windows in play mode / exported game are draggable as well. Are they? By the "title" area or by the entire window? Only on empty space?

Like to comment just before, in the "Extend to title" mode, you will have to implement your own window move because it's impossible to tell exactly what is an empty space or not, also you could want to implement a full size draggable window if you want. It should be pretty say in GDScript to implement the window movement using mouse events.

> So, ideally the menu part on the top left could be like this:
🟦 My Super Game ▼ | Scene | Debug | Editor | Help
Using "My Super Game" as the project name, and 🟦 to represent either a Godot icon or the Project's icon.

I'm concerned with the compatibility when Godot is not in "Extend to title" mode. Also, it's not the way it works on macOS. I would like to know what others think about about.

geekley commented:
> I'm concerned with the compatibility when Godot is not in "Extend to title" mode. Also, it's not the way it works on macOS. I would like to know what others think about about.

@Hilderin Ah, I'm making this suggestion not just for this mode, but in general, even when the normal OS titlebar is there. If it's implemented, it should look like that consistently. I don't use macOS, so I don't know the implications in this case.

Hilderin commented:
@Calinou 
I changed a bit the icons to be smaller and less "chunky" (1 pixel path width instead of 2 pixels). Now that the window icons are not aligned with the buttons in Godot, having a different style was essential because it look just not aligned. The buttons are now 32x32. It's like that in Steam, I personally like the smaller buttons, it saves a couple of pixels, but I can make them the same size as the default Windows 11 size.
Also removed the background when not hover the buttons.

![image](https://github.com/user-attachments/assets/5b6e7560-f9a8-4bd6-bcbd-0b9d4f472d10)

When maximized:
![image](https://github.com/user-attachments/assets/0fc6cf10-d578-4268-bb25-052f1eb4a064)
![image](https://github.com/user-attachments/assets/66ac3574-235f-4ecb-b877-520d042836f4)

What do think?

These changed are not pushed yet, I still need to fix the full screen mode, the margin to put the menu in the corner and the distorting bug when any form of pillar/letterboxing is applied.

Hilderin commented:
Made some fixes/adjustments:
- Fixed fullscreen mode where the space for the buttons was left empty.
- Fixed issue when exiting fullscreen mode where the buttons were not visible again.
- Fixed Project Manager and Editor showing a border for a brief period during startup.
- Made the icons for Minimize, Maximize, and Restore a bit smaller.
- Adjusted the button size depending on spacing settings in the editor.
- Fixed an empty line at the bottom when maximized at startup.
- Fixed Right-To-Left layout.
- Fixed stretch mode.

Default theme:
![image](https://github.com/user-attachments/assets/37c3560a-ef52-4065-ae45-f9244c9ac3c5)

Compact:
![image](https://github.com/user-attachments/assets/b5b94beb-c2bc-4abc-9658-10ec24466151)

No spacing at all:
![image](https://github.com/user-attachments/assets/bc7bc694-9039-4905-8a69-8d1c3fdb95e3)

Fullscreen mode:
![image](https://github.com/user-attachments/assets/88b640c7-2729-4ecd-a53c-7c5c49cd7b44)

Right-to-Left:
![image](https://github.com/user-attachments/assets/e17b6894-09d8-43ff-8069-a6043b2f723f)


@Calinou This should fix everything you mentioned in your previous post, except for moving the main menu to the top-left corner. There are already some settings in the editor to specify spacing, and the editor uses these to define the global margin. I'm not sure I want to change that behavior in this PR?!?
![image](https://github.com/user-attachments/assets/d14ad0c5-9e4d-4338-8b5c-5bb6a3bd1a03)

Geometror reviewed (commented):
While the project manager is working for me, opening a project results in a fully transparent window on Win11.

![BlnztpsSOk](https://github.com/user-attachments/assets/3a53d56e-610e-49e2-a247-663bcdd641cc)

Geometror commented:
System info:
Godot v4.3.stable - Windows 10.0.22631 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Laptop GPU (NVIDIA; 31.0.15.4692) - 12th Gen Intel(R) Core(TM) i7-12700H (20 Threads)
Main screen: 2560x1600, 165Hz
Second screen: 1920x1080, 60Hz

The problem persist even on a fresh config (removed C:\Users[username]\AppData\Roaming\Godot folder).

I have a monitor connected to my laptop right now, which seems to cause the issue: the editor becomes visible when I disconnect the monitor.

Hilderin commented:
Thanks a lot, that is very helpfull, I'll look into it.

Mickeon commented on doc/classes/DisplayServer.xml:
```diff
@@ -1511,6 +1511,7 @@
 			<param index="0" name="window_id" type="int" default="0" />
 			<description>
 				Returns left margins ([code]x[/code]), right margins ([code]y[/code]) and height ([code]z[/code]) of the title that are safe to use (contains no buttons or other elements) when [constant WINDOW_FLAG_EXTEND_TO_TITLE] flag is set.
+				[b]Note:[/b] On Linux amd Windows, this always returns (0, 0) since the window buttons are not displayed by the OS when [constant WINDOW_FLAG_EXTEND_TO_TITLE] flag is set.
```
```suggestion
				[b]Note:[/b] On Linux and Windows, this method always returns [code]Vector2(0, 0)[/code], since the decoration buttons are not displayed by the OS when the [constant WINDOW_FLAG_EXTEND_TO_TITLE] flag is set.
```
or (I do prefer this personally)
```suggestion
				[b]Note:[/b] On Linux and Windows, this method always returns [constant Vector2.ZERO], since the decoration buttons are not displayed by the OS when the [constant WINDOW_FLAG_EXTEND_TO_TITLE] flag is set.
```

Not sure if these are called "decoration buttons" but this is the term you used in another description.

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -782,7 +782,7 @@
 		<member name="interface/editor/expand_to_title" type="bool" setter="" getter="">
 			Expanding main editor window content to the title, if supported by [DisplayServer]. See [constant DisplayServer.WINDOW_FLAG_EXTEND_TO_TITLE].
-			Specific to the macOS platform.
+			[b]Note:[/b] Only supported on macOS, Linux and Windows.
```
```suggestion
			[b]Note:[/b] This setting is implemented on macOS, Linux, and Windows.
```

Mickeon commented on doc/classes/DisplayServer.xml:
```diff
@@ -2094,7 +2095,8 @@
 			Use [method window_set_window_buttons_offset] to adjust minimize/maximize/close buttons offset.
 			Use [method window_get_safe_title_margins] to determine area under the title bar that is not covered by decorations.
-			[b]Note:[/b] This flag is implemented only on macOS.
+			[b]Note:[/b] This flag is implemented only on macOS, Linux and Windows.
```
```suggestion
			[b]Note:[/b] This flag is implemented on macOS, Linux, and Windows.
```

Mickeon commented on doc/classes/DisplayServer.xml:
```diff
@@ -1873,7 +1874,7 @@
 		</constant>
 		<constant name="FEATURE_EXTEND_TO_TITLE" value="20" enum="Feature">
-			Display server supports expanding window content to the title. See [constant WINDOW_FLAG_EXTEND_TO_TITLE]. [b]macOS[/b]
+			Display server supports expanding window content to the title. See [constant WINDOW_FLAG_EXTEND_TO_TITLE]. [b]macOS, Linux and Windows[/b]
```
Not a fan of listing them like this but it's consistent with the rest.
```suggestion
			Display server supports expanding window content to the title. See [constant WINDOW_FLAG_EXTEND_TO_TITLE]. [b]macOS, Linux, Windows[/b]
```

Mickeon commented on doc/classes/DisplayServer.xml:
```diff
@@ -2094,7 +2095,8 @@
 			Use [method window_get_safe_title_margins] to determine area under the title bar that is not covered by decorations.
-			[b]Note:[/b] This flag is implemented only on macOS.
+			[b]Note:[/b] This flag is implemented only on macOS, Linux and Windows.
+			[b]Note:[/b] On Linux and Windows, the decoration buttons (minimize, maximize and close) are rendered by Godot and use the Window theme.
```
I believe, being the name of the buttons, that these should be capitalized (?)
```suggestion
			[b]Note:[/b] On Linux and Windows, the decoration buttons (Minimize, Maximize and Close) are rendered by Godot and use the Window's theme.
```
or, because the "possessive 's" could be confused with "Windows" itself
```suggestion
			[b]Note:[/b] On Linux and Windows, the decoration buttons (Minimize, Maximize and Close) are rendered by Godot and use the theme of the Window.
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -846,7 +846,8 @@
 			Main window content is expanded to the full size of the window. Unlike a borderless window, the frame is left intact and can be used to resize the window, and the title bar is transparent, but has minimize/maximize/close buttons.
-			[b]Note:[/b] This setting is implemented only on macOS.
+			[b]Note:[/b] This setting is implemented only on macOS, Linux and Windows and when [member display/window/subwindows/embed_subwindows] is [code]false[/code].
+			[b]Note:[/b] On Linux and Windows, the decoration buttons (minimize, maximize and close) are rendered by Godot and use the Window theme.
```
Same thoughts as a prior comment

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -846,7 +846,8 @@
 		<member name="display/window/size/extend_to_title" type="bool" setter="" getter="" default="false">
 			Main window content is expanded to the full size of the window. Unlike a borderless window, the frame is left intact and can be used to resize the window, and the title bar is transparent, but has minimize/maximize/close buttons.
-			[b]Note:[/b] This setting is implemented only on macOS.
+			[b]Note:[/b] This setting is implemented only on macOS, Linux and Windows and when [member display/window/subwindows/embed_subwindows] is [code]false[/code].
```
```suggestion
			[b]Note:[/b] This setting is implemented on macOS, Linux, and Windows. It also requires [member display/window/subwindows/embed_subwindows] to be [code]false[/code].
```
But it may also be good to split these in two different notes.
```suggestion
			[b]Note:[/b] This setting is implemented on macOS, Linux, and Windows. 
			[b]Note:[/b] This setting only works if [member display/window/subwindows/embed_subwindows] is [code]false[/code].
```

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -835,8 +835,9 @@
 		<constant name="FLAG_EXTEND_TO_TITLE" value="6" enum="Flags">
 			Window content is expanded to the full size of the window. Unlike borderless window, the frame is left intact and can be used to resize the window, title bar is transparent, but have minimize/maximize/close buttons. Set with [member extend_to_title].
-			[b]Note:[/b] This flag is implemented only on macOS.
+			[b]Note:[/b] This flag is implemented only on macOS and Windows.
```
```suggestion
			[b]Note:[/b] This flag is implemented on macOS and Windows.
```

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -835,8 +835,9 @@
-			[b]Note:[/b] This flag is implemented only on macOS.
+			[b]Note:[/b] This flag is implemented only on macOS and Windows.
 			[b]Note:[/b] This flag has no effect in embedded windows.
+			[b]Note:[/b] On Windows, the window buttons (minimize, maximize and close) are not displayed.
```
Same thoughts as a prior comment

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -907,6 +908,18 @@
 	<theme_items>
+		<theme_item name="decoration_button_hover_modulate" data_type="color" type="Color" default="Color(0.95, 0.95, 0.95, 1)">
+			Window button modulation color, when the button is hovered.
+			[b]Note:[/b] Used only when extend to title is active.
```
All of these notes should mention the FLAG_EXTEND_TO_TITLE flag directly.

I also am not sure if these should be notes in the first place. Maybe? But they're not notes, or quirks. They are basically required in order for these items to be useful. But there's no exact rule on what should and shouldn't be a note, either.

Hilderin replied:
I modified all the new theme elements to removed the note and add a reference to `extend_to_title` and `buttons are rendered by Godot.`

Mickeon reviewed (commented):
I was waiting for the implementation to settle in.

Hilderin commented:
Thanks for the review @Mickeon, I made the suggested modifications the best I can.

scgm0 commented:
The project manager that extends to titles is not a good experience under Linux X11:
1. Moving the window and changing the window size will not trigger desktop effects, such as window shaking.
2. Moving window is not smooth
3. The delay in changing the window size is very high.

This pr:

https://github.com/user-attachments/assets/d04c8a6b-ebe5-4a3f-b4ee-b710f202408b

My experimental project:

https://github.com/user-attachments/assets/e8c99d41-454d-405d-9ad1-3649c9a7f069

My project code:
![图片](https://github.com/user-attachments/assets/757fa213-dccc-48c3-806a-3b613847bde1)

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -916,9 +926,21 @@
 			Vertical position offset of the close button.
 		</theme_item>
+		<theme_item name="maximize_h_offset" data_type="constant" type="int" default="36">
+			Horizontal position offset of the maximize/restore decoration button, when [member extend_to_title] is true and the button is rendered by Godot.
```
Many other descriptions in the PR would have to be adjusted as follows:
```suggestion
			Horizontal position offset of the maximize/restore decoration button, when [member extend_to_title] is [code]true[/code] and the button is rendered by Godot.
```

scgm0 commented:
Perhaps the window dragging and resizing in this pr can be replaced using the api provided by these two pr's?
https://github.com/godotengine/godot/pull/100532
https://github.com/godotengine/godot/pull/100464

scgm0 commented:
This pr's minimize button differs from the native minimize button: you need to click the taskbar icon twice to redisplay the window after minimizing, and it won't keep the window maximized, which is a bug of directly setting the window mode to `MODE_MINIMIZED`.
My desktop environment is archlinux + kde6 + wayland.
https://github.com/user-attachments/assets/c16b5f4d-1a7a-4018-869e-2b08e70e078d

https://github.com/user-attachments/assets/bf1dc5eb-eb41-4869-9947-816b10b284da

bruvzg commented:
> Perhaps the window dragging and resizing in this pr can be replaced using the api provided by these two pr's?

At least for Linux it definitely better to do so, this will add Wayland support and is an improvement over manually moving window on X11. Windows code is OK as is, but can be replaced as well, for consistency.

TheESN commented:
Though some Wayland environments (for example KDE) prefers to give titlebars themselves, so you need to give proper preference through xdg-decoration protocol (Also maybe make the default for server side titlebars for such environments if their such preference is detected)

Nallebeorn commented:
On modern Windows, is it actually neccessary to use the workaround with a borderless window? The custom titlebar buttons and dragging are missing out on some features, like the "quick snap" options in Windows 11 when right-clicking on maximize in Windows. And at least in its current state, the window dragging does not trigger snap points when at the edge of the screen and doesn't let you drag down a maximised window to unmaximise it (I don't know if https://github.com/godotengine/godot/pull/100464 might fix these).

Plenty of applications on Window have something like this but with seemingly native title bar buttons -- e.g. Chrome and lots of Electron apps, JetBrains IDEs. even Notepad in Win11 has tabs in the title bar area.

MattParkerDev commented:
Looking forward to this. Avalonia handles this well, at least on Windows, using
```C#
ExtendClientAreaToDecorationsHint = true;
ExtendClientAreaChromeHints = ExtendClientAreaChromeHints.NoChrome;
```
and `ExtendClientAreaChromeHints = ExtendClientAreaChromeHints.PreferSystemChrome;` is a managed implementation of the decorations.
